tests/testsuite/sbom.rs

```diff
@@ -0,0 +1,333 @@
+                "strip": {
+                    "deferred": "None"
+                }
+            },
+            "packages": [],
+            "features": [],
+            "rustc": {
+                "version": "[..]",
+                "wrapper": null,
+                "commit_hash": "[..]",
+                "host": "[..]"
+            }
+        }
+        "#,
+    );
+}
+
+#[cargo_test]
+fn build_sbom_using_env_var() {
+    let p = project()
+        .file("Cargo.toml", &basic_bin_manifest("foo"))
+        .file("src/foo.rs", r#"fn main() {}"#)
+        .build();
+
+    p.cargo("build -Zsbom")
+        .env("CARGO_BUILD_SBOM", "true")
+        .masquerade_as_nightly_cargo(&["sbom"])
+        .run();
+
+    let file = p.bin("foo").with_extension("cargo-sbom.json");
+    assert!(file.is_file());
+}
+
+#[cargo_test]
+fn build_sbom_project_bin_and_lib() {
+    let p = configured_project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "1.2.3"
+                authors = []
+
+                [lib]
+                crate-type = ["rlib"]
+            "#,
+        )
+        .file("src/main.rs", r#"fn main() { let _i = foo::give_five(); }"#)
+        .file("src/lib.rs", r#"pub fn give_five() -> i32 { 5 }"#)
+        .build();
+
+    p.cargo("build -Zsbom")
+        .masquerade_as_nightly_cargo(&["sbom"])
+        .run();
+
+    assert!(p.bin("foo").with_extension("cargo-sbom.json").is_file());
+    assert_eq!(
+        2,
+        p.glob(p.target_debug_dir().join("*.cargo-sbom.json"))
+            .count()
+    );
+}
+
+#[cargo_test]
+fn build_sbom_with_simple_build_script() {
+    let p = configured_project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+                build = "build.rs"
+            "#,
+        )
+        .file("src/main.rs", "#[cfg(foo)] fn main() {}")
+        .file(
+            "build.rs",
+            r#"fn main() { println!("cargo::rustc-cfg=foo"); }"#,
+        )
+        .build();
+
+    p.cargo("build -Zsbom")
+        .masquerade_as_nightly_cargo(&["sbom"])
+        .run();
+
+    let path = p.bin("foo").with_extension("cargo-sbom.json");
+    assert!(path.is_file());
+}
+
+#[cargo_test]
+fn build_sbom_with_build_dependencies() {
+    Package::new("baz", "0.1.0").publish();
+    Package::new("bar", "0.1.0")
+        .build_dep("baz", "0.1.0")
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "bar"
+                version = "0.1.0"
+                build = "build.rs"
+
+                [build-dependencies]
+                baz = "0.1.0"
+            "#,
+        )
+        .file("src/lib.rs", "pub fn bar() -> i32 { 2 }")
+        .file(
+            "build.rs",
+            r#"fn main() { println!("cargo::rustc-cfg=foo"); }"#,
```

Comment:
You may need to add one more build script instruction `cargo::rustc-check-cfg=cfg(has_foo)` to avoid `unexpected_cfg` lint error.

See <https://blog.rust-lang.org/2024/05/06/check-cfg.html>.